### ISSUE TITLE
Add `pretrain_epocs` to VilLain

### ISSUE THREAD

**Issue by tizianocitro** [hypernetwork-research-group/hypertorch#345]
### Is your feature request related to a problem? Please describe.

  `VilLainNcModule` and `VilLainHlpModule` currently train with joint supervised and VilLain self-supervised losses:

  ```python
  # NC
  loss = nc_loss + villain_loss_weight * villain_loss

  # HLP
  loss = hlp_loss + villain_loss_weight * villain_loss
  ```

This preserves the current HyperTorch pipeline behavior, but it does not match the reference VilLain workflow exactly. In the reference implementation, VilLain is trained as a self-supervised embedding generator using only local/global VilLain losses, then embeddings are generated under torch.no_grad() for downstream use.

  Reference sources:

  - https://raw.githubusercontent.com/geon0325/VilLain/main/code/main.py
  - https://raw.githubusercontent.com/geon0325/VilLain/main/code/model.py
  - https://raw.githubusercontent.com/geon0325/VilLain/main/code/run.sh
  - https://github.com/geon0325/VilLain

  This distinction matters because VilLain.node_embeddings(...) and VilLain.hyperedge_embeddings(...) are inference-style embedding generation paths. The supervised NC/HLP losses train the downstream classifier or decoder, while VilLain’s self-supervised objective trains the encoder.

  ### Describe the solution you would like

  Add an optional staged training mode to both VilLainNcModule and VilLainHlpModule, while keeping current joint-loss behavior as the default.

  Requested behavior:

  - Add pretrain_epochs: int | None = None to both modules.
  - If pretrain_epochs is None, keep current joint-loss behavior unchanged.
  - If pretrain_epochs is set:
      - For epochs < pretrain_epochs, train only the VilLain encoder with VilLain.loss(...).
      - For epochs >= pretrain_epochs, freeze VilLain and train only the downstream head:
          - NC: train the MLP classifier on generated node embeddings.
          - HLP: train the SLP decoder on generated node or hyperedge embeddings.

      - Validation/test after pretraining should use generated embeddings and normal task loss/metrics.

  - Keep hypertorch/models/villain.py unchanged, including torch.no_grad() in embedding generation.
  - Update examples/node_classification/villain.py and examples/hyperlink_prediction/villain.py to use staged mode where appropriate.
  - Update docs/docstrings to explain the difference between default joint mode and optional staged mode.

  ### Describe alternatives you've considered

  One alternative is to remove torch.no_grad() from VilLain embedding generation so supervised NC/HLP gradients update the VilLain embedding table directly. That would make the pipelines end-to-end supervised, but it would diverge from the reference implementation and could be expensive because embedding generation runs multiple propagation steps.

  Another alternative is to replace current behavior entirely with staged training. That would be more reference-faithful, but it would break existing public behavior for users already relying on the joint-loss pipelines.

  The preferred approach is opt-in staged training via pretrain_epochs, with current behavior preserved when it is not set.

  ### Additional context

  Implementation and testing expectations:

  - Preserve current behavior when pretrain_epochs is None.
  - Treat pretrain_epochs=0 as “skip pretraining and train only the downstream classifier/decoder on generated VilLain embeddings.”
  - Add integration coverage for both VilLainNcModule and VilLainHlpModule:
      - default joint-loss mode
      - staged mode with small values such as pretrain_epochs=1, max_epochs=2, and generation_steps=4

  - Run:
      - make si-test T=nc/villain_nc_integration_test.py
      - make si-test T=hlp/villain_hlp_integration_test.py
      - uv run python -m py_compile examples/node_classification/villain.py examples/hyperlink_prediction/villain.py
      - make check
